### PR TITLE
REF: _cython_agg_blocks follow patterns similar to _apply_blockwise

### PR DESCRIPTION
**Pull request by jbrockmendel**
Follows #35535 

This isn't yet identical to apply_blockwise, because both this and apply_blockwise have some odd behaviors that need to get standardized before they get combined into a BlockManager method, but this is a move in that direction.

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -1107,21 +1103,35 @@ def cast_result_block(result, block: "Block", how: str) -> "Block":
                         # is a lie. To keep the code-path for the typical non-split case
                         # clean, we choose to clean up this mess later on.
                         assert len(locs) == result.shape[1]
+                        new_blocks = []
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1106]: can you type this

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -1059,8 +1058,7 @@ def cast_result_block(result, block: "Block", how: str) -> "Block":
             agg_block: Block = block.make_block(result)
             return agg_block
 
-        for block in data.blocks:
-            # Avoid inheriting result from earlier in the loop
+        def blk_func(block):
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1061]: can you type this

File: pandas/core/groupby/generic.py [hunk 3]
```diff
@@ -1107,21 +1103,35 @@ def cast_result_block(result, block: "Block", how: str) -> "Block":
...
-                        agg_blocks.append(agg_block)
+                        new_blocks = [agg_block]
             else:
                 agg_block = cast_result_block(result, block, how)
-                new_items.append(locs)
-                agg_blocks.append(agg_block)
+                new_blocks = [agg_block]
+            return new_blocks
+
+        skipped = []
```
Comment (jreback) [inline pandas/core/groupby/generic.py:1121]: can you type this

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/core/groupby/generic.py]: done

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (jreback) [reply pandas/core/groupby/generic.py:1108]: can you type this

Comment (jreback) [reply pandas/core/groupby/generic.py:1108]: and move this to a higher scope (as used in both parts)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/generic.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/groupby/generic.py:1108]: We only append to it in this scope, everywhere else we set it directly

Comment (jreback) [reply pandas/core/groupby/generic.py:1108]: this is used in 3 blackes (sure you don't append and instead directly assign in the other scopes), but it is the same name. I would simply define it once, then append for each (or leave the assign for the others is ok i guess); it just is much easier to grok the fact that we are defining the same variable.

Comment (jbrockmendel) [reply pandas/core/groupby/generic.py]: done+green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks

**Pull request merged**